### ISSUE TITLE
[clang-tidy] `misc-const-correctness` false positive when calling a function with default lambda argument

### ISSUE THREAD

**Issue by BobIsOnFire** [llvm/llvm-project#97701]
Hi! Here's minimal repro:

```c++
class A {};
void poison(void(*f)() = []{}) {}
void foo() {
    A a;
    a = {};
    poison();
}
```

I get following warning (godbolt: https://godbolt.org/z/cedY7o5ev):

```
<source>:4:5: warning: variable 'a' of type 'A' can be declared 'const' [misc-const-correctness]
    4 |     A a;
      |     ^
      |       const 
1 warning generated.
```

Even though I assign to `a` on the next line. Something is wrong with calling a function with a function pointer type (or `std::function`) that defaults to a lambda, it completely messes up the check -- note that it does not even use `a` in any way. Following changes would remove the false-positive:

* Removing the argument from `poison` function: https://godbolt.org/z/qvcz4jqE1
* Replacing lambda with `nullptr`: https://godbolt.org/z/cTTW84x15
* Explicitly passing a lambda to the call: https://godbolt.org/z/YEaWjY15r

Thanks in advance! Let me know if any other information is needed.

**Comment by pkaff** [comment]
The compiler used in the godbolt link, clang 18.1, does not seem to give me the warning currently. Is this still reproducible in godbolt? As mentioned in #127494 I specifically encountered this on clang for MacOS.

**Comment by BobIsOnFire** [comment]
@pkaff still reproduces locally for me on clang 19.1.7. It doesn't reproduce on godbolt anymore because godbolt always uses trunk version of clang-tidy, regardless of compiler selected.

I've bisected history between 19.1.7 and trunk, and found that my issue was fixed in #117474 (which will be included in version 20). Cherry-picked changes locally on top of 19.1.7, and the warning is now gone.
